### ISSUE TITLE
Crash when using VSCode without trailing newline

### ISSUE THREAD

**Issue by Alluet** [zigtools/zls#129]
Possibly due to a bug in Visual Studio Code, ZLS crashes when writing new declarations (`struct`, `var`, `fn`, etc.) My editor is configured to include a trailing newline on save, but until I save this implicit newline is not actually present. This requires me to manually add padding newlines when I'd otherwise let the editor do for me.

This is with latest master build of Zig and ZLS, running in WSL 1 or 2, with Visual Studio Code connected via the "remote" feature, in case you cannot reproduce this under other conditions.

<details>
<summary>Complete Zig Error Trace</summary>

```
error: InvalidParams
[debug-main] Took 905ms to process method textDocument/didOpen

[debug-main] Took 0ms to process method textDocument/completion

/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zls/src/main.zig:1268:5: 0x281e2a in extractErr (zls)
    return error.HackDone;
    ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zls/src/types.zig:153:13: 0x2928df in types.TextDocument.positionToIndex (zls)
            return error.InvalidParams;
            ^
/home/alluet/workspace/zig-ext/zls/src/main.zig:1089:27: 0x283ff4 in completionHandler (zls)
        const pos_index = try handle.document.positionToIndex(req.params.position);
                          ^
/home/alluet/workspace/zig-ext/zls/src/main.zig:1267:19: 0x28466f in extractErr (zls)
    val catch |e| return e;
                  ^
/home/alluet/workspace/zig-ext/zls/src/main.zig:1340:17: 0x273968 in processJsonRpc (zls)
        else => return err,
                ^
/home/alluet/workspace/zig-ext/zls/src/main.zig:1511:9: 0x2693a5 in main (zls)
        try processJsonRpc(&arena, &json_parser, buf, config);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zls/src/main.zig:1268:5: 0x27f86a in extractErr (zls)
    return error.HackDone;
    ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:48:9: 0x262a1b in std.mem.Allocator.callResizeFn (zls)
        return self.resizeFn(self, buf, new_len, len_align);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/os.zig:1145:23: 0x26277b in std.os.openatZ (zls)
            ENOENT => return error.FileNotFound,
                      ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/fs.zig:708:13: 0x24935a in std.fs.Dir.openFileZ (zls)
            try os.openatZ(self.fd, sub_path, os_flags, 0);
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/fs.zig:648:9: 0x24733f in std.fs.Dir.openFile (zls)
        return self.openFileZ(&path_c, flags);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/fs.zig:1220:20: 0x297f4d in std.fs.Dir.readFileAllocOptions (zls)
        var file = try self.openFile(file_path, .{});
                   ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/fs.zig:1206:9: 0x27725f in std.fs.Dir.readFileAlloc (zls)
        return self.readFileAllocOptions(allocator, file_path, max_bytes, @alignOf(u8), null);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/os.zig:1145:23: 0x26277b in std.os.openatZ (zls)
            ENOENT => return error.FileNotFound,
                      ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/fs.zig:708:13: 0x24935a in std.fs.Dir.openFileZ (zls)
            try os.openatZ(self.fd, sub_path, os_flags, 0);
            ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/fs.zig:648:9: 0x24733f in std.fs.Dir.openFile (zls)
        return self.openFileZ(&path_c, flags);
        ^
/home/alluet/workspace/zig-ext/zig-linux-x86_64-0.6.0+e120b07a5/lib/zig/std/mem.zig:54:13: 0x26a5e6 in std.mem.Allocator.noResize (zls)
            return error.OutOfMemory;
            ^
[Info  - 11:31:03 PM] Connection to server got closed. Server will restart.
[Error - 11:31:03 PM] Request textDocument/completion failed.
```

</details>

**Comment by watzon** [comment]
Can confirm

**Comment by alexnask** [comment]
Thanks for reporting, I will try to repro and fix this soon.

**Comment by alexnask** [comment]
This was fixed in e54a6d2522f04ca2f0feb90031c0a2d5c12225d9, thanks again for reporting.
